src/uct/ib/ud/base/ud_ep.c

```diff
@@ -216,12 +216,31 @@ static void uct_ud_ep_purge_outstanding(uct_ud_ep_t *ep)
     uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
     uct_ud_ctl_desc_t *cdesc;
     ucs_queue_iter_t iter;
+    khint_t it;
 
     ucs_trace_func("ep=%p", ep);
 
-    ucs_queue_for_each_safe(cdesc, iter, &iface->tx.outstanding_q, queue) {
+    if (uct_ib_iface_device(&iface->super)->ordered_send_comp) {
+        ucs_queue_for_each_safe(cdesc, iter, &iface->tx.outstanding.queue,
+                                queue) {
+            if (cdesc->ep == ep) {
+                ucs_queue_del_iter(&iface->tx.outstanding.queue, iter);
+                uct_ud_iface_ctl_skb_complete(iface, cdesc, 0);
+            }
+        }
+
+        return;
+    }
+
+    for (it = kh_begin(&iface->tx.outstanding.map);
```

Comment:
can use `kh_foreach_value` instead

Reply:
we cannot since we need to access `it` to potentially remove the element.

Reply:
1. I believe so as `kh_del()` is only marking it deleted
2. fixed `it` -> `khiter` in .h and .c

```C
382     SCOPE void kh_del_##name(kh_##name##_t *h, khint_t x)               \
383     {                                                                   \
384         if (x != h->n_buckets && !__ac_iseither(h->flags, x)) {         \
385             __ac_set_isdel_true(h->flags, x);                           \
386             --h->size;                                                  \
387         }                                                               \
388     }
```